README
NAME
    Soar::WM - Traverse Soar working memory dumps

VERSION
    version 0.03

SYNOPSIS
      use Soar::WM qw(wm_root_from_file);
      my $root = wm_root_from_file('/path/to/wme/dump');
      print $root->id; #probably prints S1
  
      #or
      my $wm = Soar::WM->new(text => '(S1 ^foo bar)
      (S2 ^boo far)');
      my $wme = $wm->get_wme('s2');

DESCRIPTION
    This module represents Soar's working memory. It can be used for
    traversing and manipulating WME dumps generated by Soar.

NAME
    Soar::WM - Perl extension for representing Soar working memory given a
    WME dump file

METHODS
  "new"
    Creates a new instance of Soar::WM. The arguments to this method are the
    same as those to wm_root.

  "get_wme"
    Argument: string working memory element ID ('S1', 'Z33', etc.); since
    WME ID's are always uppercase, this method is case insensitive.

    Returns a Soar::WM::Element instance representing the given ID.

  "wm_root_from_file"
    This is a shortcut for:

     wm_root(file=>$arg)

    It's single argument is the path to a WME dump file, or an opened file
    handle for one. It returns a Soar::WM::Element object representing the
    root of the given WME dump.

  "wm_root"
    This function reads in a Soar WME dump and returnes a Soar::WM::Element
    representing its root. It takes a named argument, file or text. Using
    "wm_root(file="path)> or "wm_root(file="$fileGlob)>, you can create an
    object given the path to a WME dump file.

    Using "wm_root(text="'(S1 ^foo bar)')>, you can create an object using a
    given WME dump text. If neither argument is specified, this function
    will wait for input from standard in.

SEE ALSO
    The homepage for the Soar cognitive architecture is here:
    <http://sitemaker.umich.edu/soar/home>.

AUTHOR
    Nathan Glenn <garfieldnate@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Nathan Glenn.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Soar-WM

0.03      2012-12-18 02:22:37 UTC
	Fixed some documentation, and added the children() method to Soar::WM::Element.
	
0.02      2012-12-01 00:23:58 UTC
    Fixed some documentation, as well as adding a new() method and a get_wme() method.

0.01      2012-11-29 22:08:22 UTC
	First release. Uses code written by Deryle Lonsdale to slurp a WME dump.

Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Data::Dumper" => 0,
    "Data::Section::Simple" => 0,
    "File::Find" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Module::Build" => "0.3601",
    "Test::Deep" => 0,
    "Test::More" => "0.88",
    "Test::NoWarnings" => 0,
    "Test::Warn" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Traverse Soar working memory dumps\r",
  "dist_author" => [
    "Nathan Glenn <garfieldnate\@gmail.com>"
  ],
  "dist_name" => "Soar-WM",
  "dist_version" => "0.03",
  "license" => "perl",
  "module_name" => "Soar::WM",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Exporter" => 0,
    "autodie" => 0,
    "base" => 0,
    "perl" => "5.010",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Soar/WM.pm
#
# This file is part of Soar-WM
#
# This software is copyright (c) 2012 by Nathan Glenn.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
# ABSTRACT: Traverse Soar working memory dumps
package Soar::WM;
use strict;
use warnings;

use Soar::WM::Slurp qw(read_wm);
use Soar::WM::Element;

use Carp;
use base 'Exporter';
our @EXPORT_OK = qw( wm_root_from_file wm_root );

our $VERSION = '0.03'; # VERSION

print 'root is ' . __PACKAGE__->wm_root_from_file( $ARGV[0] )->id unless caller;

sub new {
    my ($class, @slurp_args) = @_;
    my ( $wme_hash, $root_name ) = read_wm(@slurp_args);
    my $wm = bless $wme_hash, $class;
	$wm->{'#root_name'} = $root_name;
	return $wm;
}

sub get_wme {
	my ($self, $id) = @_;
	return Soar::WM::Element->new($self, uc $id);
}

#return a Soar::WM::Element object for the root
sub wm_root_from_file {
    my ($file) = @_;
    if ( !$file ) {
        carp 'missing file name argument';
        return;
    }

    return wm_root( file => $file );

}

#args should be file=>xx or text=>xx.
sub wm_root {
    my @slurp_args = @_;
    my ( $wme_hash, $root_name ) = read_wm(@slurp_args);

    my $wm = bless $wme_hash, __PACKAGE__;
    return Soar::WM::Element->new( $wm, $root_name );
}



1;

__END__

=pod

=head1 NAME

Soar::WM - Traverse Soar working memory dumps

=head1 VERSION

version 0.03

=head1 SYNOPSIS

  use Soar::WM qw(wm_root_from_file);
  my $root = wm_root_from_file('/path/to/wme/dump');
  print $root->id; #probably prints S1
  
  #or
  my $wm = Soar::WM->new(text => '(S1 ^foo bar)
  (S2 ^boo far)');
  my $wme = $wm->get_wme('s2');

=head1 DESCRIPTION

This module represents Soar's working memory. It can be used for traversing and manipulating WME dumps
generated by Soar.

=head1 NAME

Soar::WM - Perl extension for representing Soar working memory given a WME dump file

=head1 METHODS

=head2 C<new>

Creates a new instance of Soar::WM. The arguments to this method are the same as those to wm_root.

=head2 C<get_wme>

Argument: string working memory element ID ('S1', 'Z33', etc.); since WME ID's are always uppercase, this method is
case insensitive.

Returns a L<Soar::WM::Element> instance representing the given ID. 

=head2 C<wm_root_from_file>

This is a shortcut for:

 wm_root(file=>$arg)

It's single argument is the path to a WME dump file, or an opened file handle for one. It returns a L<Soar::WM::Element> object representing the root
of the given WME dump.

=head2 C<wm_root>

This function reads in a Soar WME dump and returnes a L<Soar::WM::Element> representing its root.
It takes a named argument, file or text. Using C<wm_root(file=>path)> or C<wm_root(file=>$fileGlob)>, you can create an object given the path to a WME dump file.

Using C<wm_root(text=>'(S1 ^foo bar)')>, you can create an object using a given WME dump text. 
If neither argument is specified, this function will wait for input from standard in.

=head1 SEE ALSO

The homepage for the Soar cognitive architecture is here: L<http://sitemaker.umich.edu/soar/home>.

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Nathan Glenn.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Soar/WM/Slurp.pm
#
# This file is part of Soar-WM
#
# This software is copyright (c) 2012 by Nathan Glenn.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Soar::WM::Slurp;

use strict;
use warnings;
use 5.010;
use autodie;
use Carp;

use base qw(Exporter);
our @EXPORT_OK = qw(read_wm_file read_wm);

our $VERSION = '0.03'; # VERSION
# ABSTRACT: Read and parse Soar working memory dumps

say Dump read_wm( file => $ARGV[0] ) unless caller;

sub read_wm_file {
    my ($file) = @_;
    return read_wm( file => $file );
}

#structure will be:
# return_val->{$wme} = { $attr=>[@values]}
# {'root_wme'} = 'S1' or some such
#parse a WME dump file and create a WM object; return the WM hash and the name of the root WME.
sub read_wm {    ## no critic (RequireArgUnpacking)
    my %args = (
        text => undef,
        file => undef,
        @_
    );
    my $fh;
    if ( $args{text} ) {
        $fh = _get_fh_from_string( $args{text} );
    }
    elsif ( $args{file} ) {
        $fh = _get_fh( $args{file} );
    }
    else {
        $fh = \*STDIN;
        print "Reading WME dump from standard in.\n";
    }

    #control variables
    my ( $hasOpenParen, $hasCloseParen );

    #keep track of results/return value
    my ( $root_wme, %wme_hash );
    while ( my $inline = <$fh> ) {
        chomp $inline;
        next if $inline eq '';
        my $line = "";

        #note: do we need $hasOpenParen?
        $hasOpenParen  = ( $inline =~ /^\s*\(/ );
        $hasCloseParen = ( $inline =~ /\)\s*$/ );

        #read entire space between parentheses
        while ( $hasOpenParen && !($hasCloseParen) ) {
            chomp $inline;
            $line .= $inline;
            $inline = <$fh>;

            #if this line of the WME dump is incomplete, ignore it.
            if ( !$inline ) {
                $inline = '';
                $line   = '';
                last;
            }
            $hasCloseParen = ( $inline =~ /\)\s*$/ );
        }
        $line .= $inline;
        if ($line) {

            #separate wme and everything else [(<wme> ^the rest...)]
            my ( $wme, $rest ) = split " ", $line, 2;

            # initiate the record
            my $rec = {};

            # hash each of the attr/val pairs
            my @attVals = split /\^/, $rest;

            #if line were 'S16 ^foo bar ^baz biff', then @attvals
            #now contains ['S16', 'foo bar', 'baz biff']

            #get rid of the WME ID
            shift @attVals;

            foreach my $attVal (@attVals) {
                my ( $attr, $val ) = split " ", $attVal;
                if ( !length($attr) ) {    #note: would this ever happen?
                    next;
                }

                #get rid of final parenthesis
                $val =~ s/\)$//;

                # store attr/val association in the record
                push @{ $rec->{"$attr"} }, $val;
            }

            #strip opening parenthesis
            $wme =~ s/^\(//;

            # $rec->{'#wmeval'} = $wme;

            #rootwme is S1, or similar
            $root_wme = $wme unless $root_wme;

            # add the record to the wme hash
            $wme_hash{$wme} = $rec;
        }
    }
    close $fh;
    return \%wme_hash, $root_wme;
}

sub _get_fh_from_string {
    my ($text) = @_;
    open my $sh, '<', \$text;
    return $sh;
}

sub _get_fh {
    my ($name) = @_;
    return $name if ref $name eq 'GLOB';
    open my $fh, '<', $name;
    return $fh;
}

1;

__END__

=pod

=head1 NAME

Soar::WM::Slurp - Read and parse Soar working memory dumps

=head1 VERSION

version 0.03

=head1 SYNOPSIS

  use Soar::WM::Slurp qw(read_wm);
  use Data::Dumper;
  my ($WM_hash, $root_name) = read_wm(file => 'path/to/wme/dump/file');
  print 'root is ' . $root_name;
  print Dumper($WM_hash);

=head1 NAME

Soar::WM::Slurp - Perl extension for slurping Soar WME dump files.

=head1 DESCRIPTION
This module can be used to read in a Soar WME dump file. It exports one function, read_wm, which reads a WME dump and returns a hash pointer representing it.

=head1 METHODS

=head2 C<read_wm_file>
A shortcut for C<read_wm( file=>$arg )>

=head2 C<read_wm>
Takes a named argument, either file => 'path/to/file', file => $fileGlob, or text => 'WME dump here'.
Returns a pointer to a hash structure representing the input WME dump, and the name of the root WME, in a list like this: ($hash, $root).

Note that any incomplete WME structures will be ignored; for example:

	(S1 ^foo bar ^baz boo ^link S2)
	(S2 ^faz far ^boo baz ^fuzz buzz

The second line in the above text will be ignored. Although some of the structure there is apparent, accepting incomplete structures would require much more 
error and input checking. WME dumps are normally complete, so this should not be a problem.

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Nathan Glenn.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Soar/WM/Element.pm
#
# This file is part of Soar-WM
#
# This software is copyright (c) 2012 by Nathan Glenn.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Soar::WM::Element;
use strict;
use warnings;

our $VERSION = '0.03'; # VERSION
# ABSTRACT: Work with Soar working memory elements

use Carp;

sub new {
    my ( $class, $wm, $id ) = @_;
    if ( !exists $wm->{$id} ) {
        carp 'Given ID doesn\'t exist in given working memory';
        return;
    }

    my $self = bless {
        wm   => $wm,
        id   => $id,
        node => $wm->{$id},
    }, $class;
    return $self;
}

sub id {
    my ($self) = @_;
    return $self->{id};
}

sub atts {
    my ($self) = @_;
    my @atts = keys %{ $self->{node} };
    return \@atts;
}

sub vals {
    my ( $self, $query ) = @_;
    if ( !$query ) {
        carp 'missing argument attribute name';
        return;
    }
    my @values = @{ $self->{node}->{$query} };
	
    #find ones that are links and change them into WME instances
    for ( 0 .. $#values ) {
        if ( exists $self->{wm}->{ $values[$_] } ) {
            $values[$_] = __PACKAGE__->new( $self->{wm}, $values[$_] );
        }
    }
    return \@values;
}

sub children {
    my ($self , %args) = @_;
	
	my @children;
	for my $key ( keys %{$self->{node}} ){
		push @children, @{ $self->{node}->{$key} };
	}
	
    #find ones that are links and change them into WME instances
    for ( 0 .. $#children ) {
        if ( exists $self->{wm}->{ $children[$_] } ) {
            $children[$_] = __PACKAGE__->new( $self->{wm}, $children[$_] );
        }
    }
	my $retVal;
	if($args{links_only}){
		my @links = grep {ref($_) eq 'Soar::WM::Element'} @children;
		$retVal = \@links;
	}else{
		$retVal = \@children;
	}
    return $retVal;
}

sub first_val {
    my ( $self, $query ) = @_;
    if ( !$query ) {
        carp 'missing argument attribute name';
        return;
    }

    # grab only the first value
    my $value = ${ $self->{node}->{$query} }[0];

	if(not defined $value){
		return;
	}
	
    #if value is a link, change it into a WME instance
    if ( exists $self->{wm}->{$value} ) {
        $value = __PACKAGE__->new( $self->{wm}, $value );
    }
    return $value;
}

sub num_links {
    my ($self) = @_;
    my $count = 0;

    #iterate values of each attribute; a child will have its own entry in WM
    for my $att ( @{ $self->atts } ) {
        for my $val ( @{ $self->{node}->{$att} } ) {
            $count++
              if ( exists $self->{wm}->{$val} );
        }
    }
    return $count;
}

1;

__END__

=pod

=head1 NAME

Soar::WM::Element - Work with Soar working memory elements

=head1 VERSION

version 0.03

=head1 SYNOPSIS

 use Soar::WM qw(wm_root);
 my $root = wm_root(<<ENDWM);
 (S1 ^foo bar ^foo buzz ^baz boo ^link S2 ^link S3)
 (S2 ^faz far 
    ^boo baz
    ^fuzz buzz)
 (S3 ^junk foo)
 ENDWM
 print $root->id; # 'S1'
 my $val = $root->first_val('link');
 print $val->id; # 'S2'
 print $val->first_val('faz'); # 'far'

=head1 DESCRIPTION

This module allows one to traverse working memory by accessing attributes and values of a single element at a time.

=head1 NAME

Soar::WM::Element - Perl extension for representing Soar working memory elements.

=head1 METHODS

=head2 C<new>

Creates a new instance of Soar::WM::Element. There are two required arguments:
1. An instance of L<Soar::WM> which is to contain this element
2. The WME ID of this element.
If the given ID does not exist in the given working memory, this method will croak.

=head2 C<id>

Returns the WME ID of the this element ('S1', 'W3', etc.)

=head2 C<atts>

Returns an array pointer containing the attributes present in this element.

=head2 C<vals>

Takes one required argument, an attribute name, and returns an array pointer containing all of the values of the given attribute
for this element. Any values that are names of other working memory elements will be blessed as new Soar::WM::Elements.

=head2 C<children>

The same as vals, but returns the value of every attribute as an array pointer. The optional parameter 'links_only => 1' will cause
the method to only return children which are links to other Soar::WM::Elements.

=head2 C<first_val>

Takes one required argument, an attribute name, and returns the first value of the given attribute for this element. 
If the values is the names of another working memory element, it will be blessed as a new Soar::WM::Element.

=head2 C<num_links>

Returns the number of values in all of this element's attributes which are names of other existing working memory elements.

=head1 SEE ALSO

The homepage for the Soar cognitive architecture is here: L<http://sitemaker.umich.edu/soar/home>.

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Nathan Glenn.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Traverse Soar working memory dumps\r",
  "AUTHOR" => "Nathan Glenn <garfieldnate\@gmail.com>",
  "BUILD_REQUIRES" => {
    "Data::Dumper" => 0,
    "Data::Section::Simple" => 0,
    "File::Find" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Module::Build" => "0.3601",
    "Test::Deep" => 0,
    "Test::More" => "0.88",
    "Test::NoWarnings" => 0,
    "Test::Warn" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Soar-WM",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Soar::WM",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Exporter" => 0,
    "autodie" => 0,
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.03",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



